### ISSUE TITLE
remove code that uses `posix_spawn()`

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#3149]
Working on the fix for issue #2980 caused me to notice the use of `posix_spawn()`. However, those uses are protected by symbol `FISH_USE_POSIX_SPAWN` which is never set AFAICT. That code was added four years ago (see issue #11). I want to remove that code because a) it isn't being used, and b) isn't compatible with my fix for issue #2980. The presence of `posix_spawn()` in the code probably explains why fish does things like call `tcsetpgrp()` in both the parent and child process and suppresses error messages in the parent process. Ugh!

A fish shell should never be so large that the cost of `fork()` is unreasonable (most of the cost is due to copying the MMU page tables or equivalent which scales with the size of the process). So the optimization provided by `posix_spawn()` should never be a factor for fish. Also, a bit of googling suggests that many implementations are nothing more than fork/exec wrappers in user space rather than a more efficient kernel implementation.

@ridiculousfish Any objection to removing the `posix_spawn()` code?

**Comment by zanchey** [comment]
`FISH_USE_POSIX_SPAWN` is defined to the value of `HAVE_SPAWN_H` by `postfork.h`. It certainly seems to get used on my system (`ltrace -e posix_spawn fish`).

**Comment by krader1961** [comment]
Yes, I see it now that I look closer at the grep output. It's defined indirectly:

```
src/postfork.h:#define FISH_USE_POSIX_SPAWN HAVE_SPAWN_H
```

**Comment by floam** [comment]
> ~it isn't being used~
> isn't compatible with my fix for issue #2980

Given that it is being used, and I assume that fixing #2980 correctly while being aware of posix_spawn() is easier than yarding posix_spawn() out (and easier than it was to add the feature in the first place). Do we really want to nuke it?

**Comment by krader1961** [comment]
I locally altered my fix for the external command tty modes issue by adding a `usleep(100000)` call in the `job_continue()` function. That breaks setting the tty correctly before the child process runs. The only reason this has been working is because of an implicit race condition that is normally won. Normally the parent process will continue to execute and get through that block of code before the child process is scheduled to run. But if running that parent process code is delayed then the child process gets the wrong tty environment.

This is almost certainly something that can be fixed but it makes me gnash my teeth. Neither bash or zsh use `posix_spawn()`. Why are we using it? I see no evidence that the speedup is worth the added complexity.

> posix spawn seems like the better API.

I'd like to think you're yanking my tail, @floam, but it looks like you're serious. Have you actually looked at the API and all of the fish code to use it? It's a frigging nightmare. It was clearly modeled on the equivalent MS Windows API which no one outside of Microsoft would claim is an exemplar of an elegant API.

**Comment by krader1961** [comment]
I added a call to `terminal_give_to_job()` just before calling `posix_spawn()`. That "fixes" the race and provides the external command with the correct tty modes. It also more closely approximates what the original code was doing. It also means we're doing the work of `terminal_give_to_job()` twice given how the code is currently structured to support both posix_spawn and fork()+exec(). Which makes me wonder how much speedup we're really seeing in practice from using posix_spawn(). Since the project has not been running any benchmarks there is no way to answer that question other than by guessing. And I'm guessing the speedup is quite small.

**Comment by ridiculousfish** [comment]
Whoa... posix_spawn is a significant optimization on OS X and BSDs. This is one place where we  significantly outperform bash.

**Comment by krader1961** [comment]
> This is one place where we significantly outperform bash.

How was that measured?

**Comment by floam** [comment]
> Since the project has not been running any benchmarks there is no way to answer that question other than by guessing.

I'd say the way to answer it is by doing a benchmark.

Here's a naive one. We use fork if there is a file redirection other than to /dev/null. 

```
~ $ ln -s /dev/null null
~ $ for i in (seq 1000); command echo -n "" > /dev/null ; end # posix_spawn()
~ $ echo $CMD_DURATION
1473
~ $ for i in (seq 1000); command echo -n "" ; end # posix_spawn()
~ $ echo $CMD_DURATION
1455
~ $ for i in (seq 1000); command echo -n "" > null ; end # fork()
~ $ echo $CMD_DURATION 
2183
```

**Comment by ridiculousfish** [comment]
`posix_spawn` has one race condition that I'm aware of, which is detailed in [this link](http://www.mail-archive.com/ast-developers@research.att.com/msg00719.html) discussing ksh. The other problem is that error reporting for `posix_spawn` is borked. We work around both of these by only using `posix_spawn` when 1. job control is not active, and 2. we're not redirecting with some file. In other cases we fall back to fork. See `exec.cpp:can_use_posix_spawn_for_job()` for where this logic is.

**Comment by ridiculousfish** [comment]
@floam  gives a fine example. It's a big win on OS X. Here's another, showing a comparison with bash:

```
>  cat test.sh
#!/bin/bash
for i in $(seq 1000); do
    /bin/echo "" > /dev/null
done
>  /usr/bin/time ./test.sh
        1.53 real         0.67 user         0.84 sys
>  cat test.fish
#!/usr/local/bin/fish
for i in (seq 1000)
    /bin/echo "" > /dev/null
end
>  /usr/bin/time ./test.fish
        1.13 real         0.55 user         0.47 sys
>  env fish_use_posix_spawn=0 /usr/bin/time ./test.fish
        1.72 real         0.70 user         0.93 sys
```

So our use of `posix_spawn` is the only reason we're currently able to outperform bash on OS X.

The reason we only enable this on OS X is that it's a net loss on Linux, since `posix_spawn` is implemented in terms of `fork` on Linux - unsurprising since Linux has ridiculously fast forks, thanks to the OoM killer. (I don't remember if we enable this on BSDs, but it's probably a net win there too.)

**Comment by floam** [comment]
> posix_spawn is implemented in terms of fork on Linux

It's probably implemented with `vfork` on Linux - not `fork` - so it's likely going to be faster there too. There are benchmarks [here showing posix_spawn instead of fork + exec being even more helpful on Linux than it is on OS X](https://github.com/rtomayko/posix-spawn).

**Comment by floam** [comment]
(actually, it seems to use `clone()`  on an Ubuntu machine I have access to.)

**Comment by ridiculousfish** [comment]
Nifty. It's worth investigating whether posix_spawn is now a win on Linux.

**Comment by floam** [comment]
Are you sure we aren't already? It looks like we are everywhere as long as there's a spawn.h.

**Comment by krader1961** [comment]
> I'd say the way to answer it is by doing a benchmark.  Here's a naive one.

Yes, that is an extremely naive benchmark that doesn't approximate real world usage. You did not show that the slowdown was due to the use of fork+exec. The slow down is just as likely to be due, at least in part, to the redirection. Too, the difference is 728 ms which is 0.728 ms per invocation (0.00728 seconds). Ridiculousfish provided a somewhat better example that makes it clearer that nearly all the speedup is due to the use of spawn.

The problem is that external commands run from a script such as *config.fish* (as opposed to an interactive prompt) do not have the `JOB_CONTROL` or `JOB_TERMINAL` flags set. Even when there is no I/O redirection in effect. So if those external commands try to do I/O to the tty they'll probably have problems because the tty doesn't have the correct tty modes. That's the crux of the problem in issue #2980.

Furthermore, because the tty process groups haven't been properly setup at the time the external process starts running (the race I alluded to) there is no guarantee that [ctrl-C] will kill the process. In practice that will never be noticed due to the difference between human and computer speeds. This seem to work reliably only because the current code changes the tty process group ownership within a few microseconds of the external command starting to run. But just because it isn't likely to be observed doesn't mean it isn't a race condition.

**Comment by krader1961** [comment]
Some of the issues I've noted can likely be dealt. Most notably, setting the tty modes to a sane state for external commands regardless of whether the command was run from a script (function) or interactively. But it is impossible to correctly handle the tty controlling process group when using `posix_spawn()` without race conditions.

Interestingly, the IBM implementation of the spawn() function as documented [here](https://www.ibm.com/support/knowledgecenter/SSLTBW_2.2.0/com.ibm.zos.v2r2.bpxbd00/rspawn.htm) provides a way to do this correctly. Unfortunately that solution does not appear to have been adopted by POSIX.

**Comment by krader1961** [comment]
Also, @ridiculousfish partially dealt with this back in 2012 in issue #360. Unfortunately that fix only addresses commands entered at an interactive prompt. Which, I'll grant you, is what most people will notice. But, as noted in issue #2980, it doesn't address less common situations.

There is now way to get around the fact that the `posix_spawn()` interface is borderline unusable by a shell. Which probably explains why bash and zsh don't use it.

**Comment by krader1961** [comment]
I read the [ksh discussion](http://www.mail-archive.com/ast-developers@research.att.com/msg00719.html) that @ridiculousfish linked to. it perfectly makes my point. You can't expect every external command to call `tcsetpgrp()` just because the shell decided to use `posix_spawn()`.

I'm going to modify my patch to fix the core issue even if `posix_spawn()` is used: i.e., the tty modes being wrong for external commands. It's impossible to fix the controlling terminal process group issue. Is that something that should cause us to abandon `posix_spawn()`?  The purist in me who doesn't want to deal with flakey behavior and the added code complexity says "yes". The practical engineer in me says "maybe not" given that such problems are unlikely to be noticed. I am not convinced by the sort of naive micro-benchmarks I've seen that we should use this awful API. If we had more realistic metrics that posix_spawn made a noticeable improvement to the speed of fish (to a human, not a robot) I would be more willing to accept the increased complexity and potential problems.

**Comment by ridiculousfish** [comment]
> The problem is that external commands run from a script such as config.fish (as opposed to an interactive prompt) do not have the JOB_CONTROL or JOB_TERMINAL flags set. Even when there is no I/O redirection in effect. So if those external commands try to do I/O to the tty they'll probably have problems because the tty doesn't have the correct tty modes. That's the crux of the problem in issue #2980.

Right, external commands run in non-interactive scenarios (including config.fish, completions, event handlers, fish_title, and some others) do not have JOB_CONTROL or JOB_TERMINAL set. This means fish does not create a new process group for them: they run in the process group of the shell itself. Per my understanding, other shells work this way too.

Perhaps the fix for #2980 as simple as setting and restoring the tty modes across non-interactive jobs? That makes sense to me.

> Furthermore, because the tty process groups haven't been properly setup at the time the external process starts running (the race I alluded to) there is no guarantee that [ctrl-C] will kill the process.

fish does not create a new process group for jobs without JOB_CONTROL, so there is no race here. tcsetpgrp can't race if you don't call it :)

**Comment by krader1961** [comment]
I didn't directly address this comment earlier:

> Whoa... posix_spawn is a significant optimization on OS X and BSDs. This is one place where we significantly outperform bash.

Does that performance improvement justify our behavior being wrong? The use of posix_spawn makes sense for processes such as emacs  and web browsers that are huge and don't require normal shell semantics regarding external commands. Doing better than bash on questionable micro benchmarks is a distinction that doesn't justify increased code complexity and incorrect behavior IMHO.

**Comment by ridiculousfish** [comment]
> Does that performance improvement justify our behavior being wrong? 

@krader1961 No, but to my knowledge fish has no wrong behavior related to its use of `posix_spawn`. Happy to be proven wrong.

I don't understand the antagonism towards the `posix_spawn` path. IMO it's much cleaner than fork, which has much nastier issues around multithreading. I wish we could use `posix_spawn` exclusively.

**Comment by krader1961** [comment]
> I don't understand the antagonism towards the posix_spawn path.

It's because it doesn't interact well with controlling terminal process groups. There is no reason that launching, say, `emacs` from *config.fish* should be handled any differently than if that command had been typed at an interactive prompt. If the latter puts emacs in its own process group then so should the former.

> IMO it's much cleaner than fork, which has truly nasty issues around multithreading.

It is not cleaner than fork+exec. All of the actions performed by `posix_spawn()` can safely be done in a forked child process even if the process is using threads AFAICT. Please point me to one thing that posix_spawn does that can't be done safely in a threaded program using fork. On the other hand there are quite a few things possible with the fork+exec model that aren't possible using posix_spawn. For example, changing the cwd or resource limits (i.e., `setrlimit()`) for the child process before running the external command. The only reason for the existence of this API is that it allows the kernel to provide additional optimizations with respect to memory management, the file table, and similar resources when spawning a new process. The speedup from those optimizations may justify this API. But elegance (i.e. it's "cleaner") is not one of those arguments. The posix_spawn API is a monstrosity that needs to justify its use by providing a significant (e.g., 2x) speedup for typical use cases.

Also, consider this function definition in the fish source:

```
int set_child_group(job_t *j, process_t *p, int print_errors) {
```

The only reason the `print_errors` argument exists is to paper over the fact that the function is being called in contexts where it can't perform certain actions such as changing the process group of the controlling terminal. And that's because the code is very confused about whether it should be doing so in the parent or child process. Confusion that seems to exist because of the use of `posix_spawn()`.

**Comment by krader1961** [comment]
@ridiculousfish I feel compelled to point out that I am 100% for changes that improve performance. I worked at Google for seven years where performance is always a top three concern for any change. Prior to that I was the performance expert at Sequent with respect to its flavor of UNIX (Dynix/PTX) and then IBM when they bought Sequent. I wrote a replacement for the `top` command named `top2` to help me solve performance problems on that OS. I was frequently asked to deal with customer complaints that some behavior was a few milli or even microseconds slower than they believed it should be. Using the CPU `rdtsc` opcode (read timestamp counter) was something I did frequently.

So I'm extremely sensitive to the need for better performance of any piece of software. But performance improvements have to be justified relative to increased code complexity, maintenance costs, and potential for bugs. It may be that using `posix_spawn()` in fish justifies those costs but I'm still unconvinced.

**Comment by ridiculousfish** [comment]
> It's because it doesn't interact well with controlling terminal process groups

To be precise, posix_spawn does not provide a reliable way to transfer control of the terminal to the child process group. This is a significant wart, for sure, but we should not overstate it.

> There is no reason that launching, say, emacs from config.fish should be handled any differently than if that command had been typed at an interactive prompt. If the latter puts emacs in its own process group then so should the former.

Are you arguing for a break from standard shell behavior? bash behaves the same as fish here. Run emacs from a bash prompt, emacs is run in its own process group. But run emacs from ~/.bashrc, emacs is run in bash's process group.  You can verify this with e.g. `ps axj`

Incidentally, either way, this has nothing to do with posix_spawn. posix_spawn, like fork, gives you a choice as to whether to create a new process group or not. The ugly part of posix_spawn is how it relates to the tty's controlling process group, but not to whether the process gets a new group.

> Please point me to one thing that posix_spawn does that can't be done safely in a threaded program using fork. On the other hand there are quite a few things possible with the fork+exec model that aren't possible using posix_spawn.

One example that's important to fish is error reporting. Say posix_spawn fails with ENOEXEC: we call printf() to report the error. But now, say execve() fails the same way. printf might allocate memory, which might take a lock, and that's not allowed. So we have to have our own implementation of printf (`debug_safe`).

There's also more exotic examples - `POSIX_SPAWN_START_SUSPENDED` is really cool for example. 

But anyways these are moot. We are inescapably on the hook for supporting the fork() execution path, so also supporting a posix_spawn() path can only make things more complicated. I don't dispute that at all. fish's support for `posix_spawn()` is about performance, not about simplicity.

> The posix_spawn API is a monstrosity

Well, no it's not, so there :)

I think you are massively overestimating how much complexity posix_spawn adds to fish. Its impact is entirely contained within the FISH_USE_POSIX_SPAWN checks, approximately one and a half functions. It has no tendrils or implications outside of that.

In fact, I think you are assigning some blame to posix_spawn that ought to land on fork! Case in point:

> The only reason the print_errors argument exists is to paper over the fact that the function is being called in contexts where it can't perform certain actions such as changing the process group of the controlling terminal. And that's because the code is very confused about whether it should be doing so in the parent or child process. Confusion that seems to exist because of the use of posix_spawn().

This is not so - in fact, you will find the same function in fish 1.x (see f9c63a142b8eaf9d9e4de01d0230ff63e897b97d), which predates posix_spawn!

1. Calling set_child_group() in both parent and child works around the usual fork-inflicted race. This is just standard shell practice - see ["both the parent shell and the child attempt to adjust the process group of the child process; it does not matter which succeeds first..."](http://pubs.opengroup.org/onlinepubs/009695399/functions/setpgid.html)
2. The print_errors argument exists because there is yet another fork-inflicted race: the setpgid() call may fail in the parent if the child has already exited. Since this is expected, we don't report this error.

These races are just inherent in the design of fork - nothing fish can do about them! But the posix_spawn path (via posix_spawnattr_setpgroup) avoids them. That's part of the reason it is more elegant in the abstract, though not in its realization.

I'd like to learn more about your proposed fix and how posix_spawn makes it harder?

**Comment by krader1961** [comment]
I've concluded the problem isn't the use of `posix_spawn()` per se. The primary problem is the current process management implementation needs a redesign and refactoring into more manageable pieces.

I still maintain that the `print_errors` parameter is unnecessary. It appears the specific situation it handles can be dealt with by checking whether errno == EACCESS after the failed `setpgrp()` call.

As for calling stdio functions (e.g., `fprintf()`) on the child side of a `fork()` that clearly needs to be avoided by using substitutes like `debug_safe()`. However, the current structure of the code makes it difficult to reason about whether that is being done where necessary, and only where necessary.

**Comment by krader1961** [comment]
I'm going to close this since there are bigger fish to fry. In particular refactoring all of the process management functions so that they are easier to reason about. I still suspect that the use of posix_spawn() can't be justified by the cost relative to the performance improvements. Especially in light of fixes to issues like #2007. But until that aforementioned refactoring occurs to make it easier to experiment with changes such as the one proposed by this issue the proposal is moot.
